Add Category management

ShastkivRuslan opened the pull request:
This pull request introduces new functionality for managing categories within the project, including role-based access control for users (USER) and administrators (ADMIN), full CRUD operations for categories, and integration with existing book data. Below is an overview of the implemented changes:
Key Changes:

1. Added Category entity with fields: id (PK), name (not null), and description.
2. Updated Book entity to include a Set<Category> categories field to support categorization.
3. Created CategoryRepository extending JpaRepository for database access.
4. Implemented DTOs for Category entity, enabling clean and organized data transfer.
5. Enhanced BookMapper with methods:

    - BookDto toDto(Book book);
    - Book toEntity(CreateBookRequestDto bookDto);
    - BookDtoWithoutCategoryIds toDtoWithoutCategories(Book book);

6. Created CategoryMapper with methods:

      - CategoryDto toDto(Category category);
      - Category toEntity(CategoryDto categoryDTO);
      - void updateCategoryFromDto(CategoryRequestDto requestDto, @MappingTarget Category category);

7. Developed CategoryService interface and CategoryServiceImpl with:

    - Page<CategoryDto> findAll()
    - CategoryDto getById(Long id)
    - CategoryDto save(CategoryDto categoryDto)
    - CategoryDto update(Long id, CategoryDto categoryDto)
    - void deleteById(Long id)

8. Created CategoryController with endpoints:

      - POST /api/categories - createCategory(CategoryDto categoryDto)
      - GET /api/categories - getAll()
      - GET /api/categories/{id} - getCategoryById(Long id)
      - PUT /api/categories/{id} - updateCategory(Long id, CategoryDto categoryDto)
      - DELETE /api/categories/{id} - deleteCategory(Long id)
      - GET /api/categories/{id}/books - getBooksByCategoryId(Long id)
      
9. Implemented search functionality by category IDs, allowing users to filter books based on selected categories
10. Added liquibase scripts to create 'category', 'book_category' tables, and insert test data into these tables.

This PR also includes updates to mappers, DTOs, and the Book entity to support the category relationship, allowing books to be categorized effectively in the application.

ShastkivRuslan committed:
Created category entity with required fields;

ShastkivRuslan committed:
added relation between book and category entity;

ShastkivRuslan committed:
added liquibase scripts for creating categories and books_categories tables;

ShastkivRuslan committed:
created Dto`s;

ShastkivRuslan committed:
comit;

ShastkivRuslan committed:
liquibase bugfix;

ShastkivRuslan committed:
added swagger support to CategoryController;

ShastkivRuslan committed:
fix liquibase script;

ShastkivRuslan committed:
added hibernate validation to CategoryRequestDto;

ShastkivRuslan committed:
final commit
-all tasks are completed;

ShastkivRuslan committed:
-Added the ability to search for books by category IDs.

ShastkivRuslan committed:
-refactoring;

ahoienko commented on src/main/java/ruslan/shastkiv/bookstore/controller/CategoryController.java:
```diff
@@ -0,0 +1,87 @@
+        return categoryService.getCategoryById(id);
+    }
+
+    @GetMapping("/{id}/books")
```
PreAuthorize is missing

ahoienko commented on src/main/java/ruslan/shastkiv/bookstore/service/book/BookServiceImpl.java:
```diff
@@ -56,9 +64,22 @@ public Page<BookDto> search(BookSearchParametersDto searchParametersDto, Pageabl
     }
 
+    private Set<Category> initCategories(List<Long> categoryIds) {
+        return categoryIds.stream()
+                .map(id -> categoryRepository.findById(id).orElseThrow(()
+                        -> new EntityNotFoundException("Cant find category by id: " + id)))
+                .collect(Collectors.toSet());
```
What if you want to set 1000 categories to book? You will make a DB request to get each separately this is not efficient. so please think how to such operation using 1 request only

ShastkivRuslan replied:
I agree with you, but I was also thinking about how to notify the admin which category is missing in the database.

ahoienko commented on src/main/java/ruslan/shastkiv/bookstore/service/book/BookService.java:
```diff
@@ -19,4 +20,6 @@ public interface BookService {
 
+    Page<BookDtoWithoutCategoryIds> getAllBooksByCategoryId(Long id, Pageable pageable);
 }
+
```
```suggestion
```

ahoienko commented on src/main/resources/db/changelog/db.changelog-master.yaml:
```diff
@@ -13,3 +13,14 @@ databaseChangeLog:
+      file: db/changelog/changes/11-insert-books.yaml
+  - include:
+      file: db/changelog/changes/12-insert-books_categories.yaml
+
```
```suggestion
```

ahoienko commented on src/main/resources/db/changelog/changes/12-insert-books_categories.yaml:
```diff
@@ -0,0 +1,114 @@
+        - insert:
+            tableName: books_categories
+            columns:
+              - column:
+                  name: book_id
+                  value: "1"
+              - column:
+                  name: category_id
+                  value: "1"
```
Better to use valueComputed to make sure that you made correct relation

ahoienko commented on src/main/resources/db/changelog/changes/09-create-books_categories-table.yaml:
```diff
@@ -0,0 +1,26 @@
+            tableName: books_categories
+            columns:
+              - column:
+                  name: book_id
```
Id is also unique, I assume that there is some PK constraints that includes all necessary constraints please check it

ahoienko reviewed (changes requested):
Nice, left some comments)

ShastkivRuslan committed:
Changes after mentor review;
- added @JsonIgnore annotation to the 'isDeleted' field in Category entity to exclude it from being serialized when loading BookDto with categories;
- refactored initCategories method to load all categories at once for better performance;
- added logic to check and throw an exception with the specific category id if a non-existent category id is provided during book creation:
- updated 12-insert-books_categories.yaml to use valueComputed for the book-category relation to ensure correctness of the relation.

ShastkivRuslan committed:
mvn clean package -> done;
manual testing of all endpoints -> done;

ShastkivRuslan commented:
Changes after mentor review
- added @JsonIgnore annotation to the 'isDeleted' field in Category entity to exclude it from being serialized when loading BookDto with categories;
- refactored initCategories method to load all categories at once for better performance;
- added logic to check and throw an exception with the specific category id if a non-existent category id is provided during book creation:
- updated 12-insert-books_categories.yaml to use valueComputed for the book-category relation to ensure correctness of the relation.

JJJazl commented on src/main/java/ruslan/shastkiv/bookstore/controller/CategoryController.java:
```diff
@@ -0,0 +1,88 @@
+    @PreAuthorize("hasRole('ADMIN')")
+    @Operation(summary = "Add a new category",
+            description = "This endpoint allows an admin to add a new category.")
+    public CategoryDto addCategory(@RequestBody @Valid CategoryRequestDto requestDto) {
```
Please, add an appropriate status code for POST method

JJJazl commented on src/main/java/ruslan/shastkiv/bookstore/model/Category.java:
```diff
@@ -0,0 +1,31 @@
+    private String name;
+    private String description;
+    @Column(nullable = false)
+    @JsonIgnore
+    private boolean isDeleted = false;
```
Why do we need @JsonIgnore for this field? We don't want to return a `Category ` model from our controllers as a response or part of it

JJJazl commented on src/main/java/ruslan/shastkiv/bookstore/dto/book/CreateBookRequestDto.java:
```diff
@@ -28,4 +31,7 @@ public class CreateBookRequestDto {
     private String coverImage;
+
+    @NotEmpty(message = "The book must belong to at least one category.")
+    private List<Long> categoryIds = new ArrayList<>();
```
```suggestion
    private Set<Long> categoryIds = new HashSet<>();
```

JJJazl commented on src/main/java/ruslan/shastkiv/bookstore/service/book/BookServiceImpl.java:
```diff
@@ -56,9 +65,30 @@ public Page<BookDto> search(BookSearchParametersDto searchParametersDto, Pageabl
                 -> new EntityNotFoundException("Can t find book by id: " + id));
     }
 
+    private Set<Category> initCategories(List<Long> categoryIds) {
```
```suggestion
    private Set<Category> initCategories(Set<Long> categoryIds) {
```

JJJazl commented on src/main/java/ruslan/shastkiv/bookstore/service/book/BookServiceImpl.java:
```diff
@@ -56,9 +65,30 @@ public Page<BookDto> search(BookSearchParametersDto searchParametersDto, Pageabl
+    private Set<Category> initCategories(List<Long> categoryIds) {
+        List<Category> categoriesByIds = categoryRepository.findAllById(categoryIds);
+
+        Set<Long> checkList = categoriesByIds.stream()
+                .map(Category::getId)
+                .collect(Collectors.toSet());
```
```suggestion
        Set<Long> categoryIdsInDB = categoriesByIds.stream()
                .map(Category::getId)
                .collect(Collectors.toSet());
```

ShastkivRuslan committed:
changes after mentor review:
-added @ResponseStatus annotation to addCategory controller;
-changed categoryIds type from List to Set;

okuzan commented on src/main/java/ruslan/shastkiv/bookstore/service/book/BookServiceImpl.java:
```diff
@@ -56,9 +65,30 @@ public Page<BookDto> search(BookSearchParametersDto searchParametersDto, Pageabl
+        }
+        return new HashSet<>(categoriesByIds);
+    }
+
```
```suggestion
```

okuzan commented on src/main/java/ruslan/shastkiv/bookstore/model/Category.java:
```diff
@@ -0,0 +1,31 @@
+@Table(name = "categories")
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE categories SET is_deleted = TRUE WHERE id =?")
+@SQLRestriction("is_deleted=false")
```
```suggestion
@SQLDelete(sql = "UPDATE categories SET is_deleted = TRUE WHERE id = ?")
@SQLRestriction("is_deleted = FALSE")
```

ShastkivRuslan committed:
changes after second mentor review:
-deleted empty line in BookServiceImpl;
-fixed some typo in sql queries;
